### PR TITLE
Adding subscriptions to web3-eth and input/output formatting support

### LINKED ISSUE #4046: [Rewrite] Implement Polling for web3-providers-http
`web3-providers-http` should implement a `subscribe` method from `IWeb3Provider` and use polling to provide similar functionality. 

- Consider adding argurments:
    - `interval`: sets how often (in second) a request is made
    - ~`maxNumberRequests`: sets how many total requests will be made before polling terminates itself~ Mover to #4111

### LINKED ISSUE #4093: [Rewrite] Finish integrating HTTP subscription and input/output formatters
### Remaining Tasks

- `web3-eth`
    - [x] Continue updating existing tests to account for refactors
    - [x] Init input formatter tests for each RPC method
    - [x] Init output formatter tests for each RPC method
    - [x] Init sanity check for subscriptions for each RPC method (verifies user can subscribe to each method)
    - [ ] ~Figure out how to add output formatter to event emitter for subscriptions~ Moved to #4106
    - [x] Update output formatter to handle `null` values
    - [x] Revisit TypeScript return types (changes based on provided `callOptions.returnType`)
- `web3-utils`
    - [x] Init `formatInput` test
    - [x] Init `formatOutput` test
    - [x] Init `formatRpcResultArray` test
- `web3-providers-http`
    - [x] Write test for `send` method (requires mocking)
- `web3-core-requestmanager`
    - [x] Write test for `send` and `subscribe` method that verifies `Web3RequestManager._defaultRpcOptions` works as expected

### PR DESCRIPTION
**Pull request by Spacesai1or**
### Note

The git history is a bit broken since this branch was based off an unmerged branch, and did not receive the git squash. I will attempt to comb through all the `Files changed` and make a comment on the files that are specific to this PR and need to be reviewed (I did not comment on added test files)

### Completed Tasks

- `web3-eth`
    - [x] Continue updating existing tests to account for refactors
    - [x] Init input formatter tests for each RPC method
    - [x] Init output formatter tests for each RPC method
    - [x] Init sanity check for subscriptions for each RPC method (verifies user can subscribe to each method)
    - [ ] ~Figure out how to add output formatter to event emitter for subscriptions~ Moved to #4106
    - [x] Update output formatter to handle `null` values
    - [x] Revisit TypeScript return types (changes based on provided `callOptions.returnType`)
- `web3-utils`
    - [x] Init `formatInput` test
    - [x] Init `formatOutput` test
    - [x] Init `formatRpcResultArray` test
- `web3-providers-http`
    - [x] Write test for `send` method (requires mocking)
- `web3-core-requestmanager`
    - [x] Write test for `send` and `subscribe` method that verifies `Web3RequestManager._defaultRpcOptions` works as expected

closes #4093 #4046

### CODE REVIEW

File: packages/web3-core-requestmanager/src/index.ts [hunk 1]
```diff
@@ -1,57 +1,103 @@
...
+            } else if (/^ipc:\/\//i.test(providerUrl)) {
+                return ProviderProtocol.WS;
+            }
+            return ProviderProtocol.UNKNOWN;
+        } catch (error) {
+            throw Error(`Error detecting provider protocol: ${error.message}`);
+        }
+    }
+
+    private static _defaultRpcOptions(
```
Comment (Spacesai1or) [inline packages/web3-core-requestmanager/src/index.ts:59]: Files needs review

File: packages/web3-eth/src/index.ts [hunk 2]
```diff
@@ -1,12 +1,23 @@
 import Web3RequestManager from 'web3-core-requestmanager';
-import { HttpRpcOptions } from 'web3-providers-http/types';
+import {
```
Comment (Spacesai1or) [inline packages/web3-eth/src/index.ts:2]: Files needs review

File: packages/web3-eth/types.ts [hunk 3]
```diff
@@ -1,15 +1,17 @@
-import { HttpRpcResponse } from 'web3-providers-http/types';
+import { RpcResponse } from 'web3-providers-base/types';
```
Comment (Spacesai1or) [inline packages/web3-eth/types.ts:1]: Files needs review

File: packages/web3-providers-base/types.ts [hunk 4]
```diff
@@ -1,11 +1,75 @@
+import { EventEmitter } from 'events';
```
Comment (Spacesai1or) [inline packages/web3-providers-base/types.ts:1]: Files needs review

File: packages/web3-providers-http/src/index.ts [hunk 5]
```diff
@@ -1,58 +1,142 @@
...
-        method: options.method,
-        params: options.params,
-      });
-
-      return response.data.data ? response.data.data : response.data;
-    } catch (error) {
-      throw Error(`Error sending: ${error.message}`);
-    }
-  }
+import {
```
Comment (Spacesai1or) [inline packages/web3-providers-http/src/index.ts:3]: File needs review

File: packages/web3-utils/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
+import { RpcResponseResult } from 'web3-providers-base/types';
```
Comment (Spacesai1or) [inline packages/web3-utils/src/index.ts]: File needs review

File: packages/web3-utils/types.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+export type PrefixedHexString = string;
```
Comment (Spacesai1or) [inline packages/web3-utils/types.ts:1]: File needs review

File: packages/web3-providers-base/types.ts [hunk 8]
```diff
@@ -1,11 +1,75 @@
...
+)[];
+export type RpcResponseResult =
+    | string
+    | number
+    | boolean
+    | (string | number)[]
+    | { [key: string]: string | number }
+    | BigInt
+    | null
+    | any;
```
Comment (luu-alex) [inline packages/web3-providers-base/types.ts]: i think either type check or allow the type to be any

**Review by luu-alex — APPROVED** [review]

File: packages/web3-eth/src/index.ts [hunk 9]
```diff
@@ -37,20 +85,23 @@ export default class Web3Eth {
      * @returns {Promise} Client version
```
Comment (luu-alex) [inline packages/web3-eth/src/index.ts:91]: Comments need to be updated from rpcOptions to CallOptions

File: packages/web3-providers-base/types.ts (hunk 8 above)
Comment (Spacesai1or) [reply packages/web3-providers-base/types.ts]: Going to create a separate PR to update the types for `web3-eth`, as merely removing `| any` caused an avalanche of type inconsistencies

File: packages/web3-eth/src/index.ts (hunk 9 above)
Comment (Spacesai1or) [reply packages/web3-eth/src/index.ts:91]: Going to update the doc blocks as part of the `web3-eth` type updates

File: packages/web3-providers-base/types.ts (hunk 8 above)
Comment (Spacesai1or) [reply packages/web3-providers-base/types.ts]: Fix by #4114

File: packages/web3-eth/src/index.ts (hunk 9 above)
Comment (Spacesai1or) [reply packages/web3-eth/src/index.ts:91]: Fix by #4114

**Comment by Spacesai1or** [comment]
~Blocked by #4115~

**Pull request merged**